### PR TITLE
Add new extension api

### PR DESCRIPTION
**Pull request by SpencerWBarnes**

### CODE REVIEW

File: core/extension/frontend_types.d.ts
Comment (pmconne) [inline core/extension/frontend_types.d.ts:184]: Has anyone determined how/if `instanceof` checks can be made to work when we're exporting only the types?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (pmconne) [inline core/frontend/src/extension/ExtensionAdmin.ts:20]: Only 2 ways are listed.

File: rush.json
Comment (aruniverse) [inline rush.json:38]: could you add a github codeowner rule for this pkg and the test one, and assign it caleb/bill, and myabe myself?

File: test-apps/ui-test-app/package.json
Comment (aruniverse) [inline test-apps/ui-test-app/package.json:110]: until we add the extension loader in the ts files, id hold off on adding this dep

File: core/frontend/src/test/Extension.test.ts
Comment (aruniverse) [inline core/frontend/src/test/Extension.test.ts:1]: i would delete this file if we dont actually have a test

File: core/frontend/src/extension/ExtensionHost.ts
Comment (aruniverse) [inline core/frontend/src/extension/ExtensionHost.ts:22]: we need to rename this, should also probably be alpha

File: core/extension/itwin-extension.d.ts
Comment (aruniverse) [inline core/extension/itwin-extension.d.ts:6]: need to clean this up, not being used atm,

**Review by aruniverse — COMMENTED** [review]
we should hold off on merging this until we clean up and finalize our initial alpha api; especially since we dont plan on including this for the first rc

i'll finish reviewing this once i return from vacation

File: core/extension/frontend_types.d.ts
Comment (SpencerWBarnes) [reply core/extension/frontend_types.d.ts:184]: What use cases are you interested in?
If you need to check the type of a variable that is passed in and was instantiated using the class's constructor then we can expose a function that effectively wraps the instanceof method.
However, if you plan to create the variable in the extension without the constructor then the best approach may require checking all required properties. This is far from ideal since it will likely be fragile, but I have not been able to find a better way to check type while hiding a class and its methods.

Comment (pmconne) [reply core/extension/frontend_types.d.ts:184]:
```
function doSomethingWithAView(view: ViewState) {
  if (view instanceof DrawingViewState)
    // ...
}
```

Comment (SpencerWBarnes) [reply core/extension/frontend_types.d.ts:184]: Is plan for the `view` parameter to be created on our side using the DrawingViewState constructor, or is it created in an extension as an object matching the DrawingViewState type?

Comment (pmconne) [reply core/extension/frontend_types.d.ts:184]: I am only asking about types that we don't expect the extension to instantiate - the only reason they would want a real export instead of a type export is to be able to use `instanceof`.

File: core/frontend/src/extension/ExtensionHost.ts
Comment (SpencerWBarnes) [reply core/frontend/src/extension/ExtensionHost.ts:22]: Has there been any names that we liked yet?

Comment (kabentley) [reply core/frontend/src/extension/ExtensionHost.ts:22]: How 'bout just `Extension`?

File: core/extension/frontend_types.d.ts
Comment (SpencerWBarnes) [reply core/extension/frontend_types.d.ts:184]: I believe the best solution to this would be to access the object's constructor name using `obj.constructor.name` which would show the exact class used to instantiate the object. This would cover the use case you described where the object is passed in and you want to check if it is a specific subclass of that type.

Pros: lightweight, built-in without maintenance, universal
Cons: sensitive to minification, unable to identify superclasses, dependent on magic string comparisons

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (SpencerWBarnes) [reply core/frontend/src/extension/ExtensionAdmin.ts]: Changed comment to say 2 ways instead

File: rush.json
Comment (SpencerWBarnes) [reply rush.json:38]: Added all three of you to the extension package and added Bill (Caleb was already added) to the test folder

File: test-apps/ui-test-app/package.json
Comment (SpencerWBarnes) [reply test-apps/ui-test-app/package.json]: Removed the dependency

File: core/frontend/src/test/Extension.test.ts
Comment (SpencerWBarnes) [reply core/frontend/src/test/Extension.test.ts]: Removed.

File: core/extension/frontend_types.d.ts
Comment (pmconne) [reply core/extension/frontend_types.d.ts:175]: That's not generally going to suffice. We'll have to revisit this sometime.

**Comment by SpencerWBarnes** [comment]
Items we would like to investigate but may not be needed for initial release:
- DisplayStyleSettings
- Possible removal of BeButtonEvent
- Method of exposing ToolSettings in a readonly manner

Items for future development:
- Using an `@extension` and `@public` or `@preview` decorators with extract-api to generate extension's exposed types and reals
- Using extract-api to validate that types, functions, or classes that are missing from the extension api
- Replacement for `typeof` operation
- Changes to handle additional TileGraphicProviders

**Comment by pmconne** [comment]
> * DisplayStyleSettings

Please don't delay 3.0 any further for this.

> * Replacement for `typeof` operation

Do you mean `instanceof`?

File: core/frontend/src/IModelApp.ts
Comment (pmconne) [inline core/frontend/src/IModelApp.ts:208]: > Event raised ***after*** IModelApp is finished starting up.
> on***Before***Startup

It's `@internal` so ignore if you like, but it should be called on***After***Startup.

**Review by pmconne — COMMENTED** [review]
docs/learning/ articles and example Extensions will be produced later?

**Comment by pmconne** [comment]
@aruniverse please revisit your comments.

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (pmconne) [inline core/frontend/src/extension/ExtensionAdmin.ts:14]: > 3 goals

There are either 2 or 4 goals listed.

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (pmconne) [inline core/frontend/src/extension/ExtensionAdmin.ts:18]: > Download the needed files from

from where?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (pmconne) [inline core/frontend/src/extension/ExtensionAdmin.ts:38]: Delete commented-out code.

File: core/frontend/src/extension/Extension.ts
Comment (pmconne) [inline core/frontend/src/extension/Extension.ts:54]: Delete commented-out code.

File: core/frontend/src/extension/ExtensionHost.ts
Comment (pmconne) [reply core/frontend/src/extension/ExtensionHost.ts:22]: ExtensionHost seems fine to me.

File: core/frontend/src/extension/ExtensionLoader.ts
Comment (pmconne) [inline core/frontend/src/extension/ExtensionLoader.ts:35]: Delete commented-out code.

File: core/frontend/src/extension/extensions.ts
Comment (pmconne) [inline core/frontend/src/extension/extensions.ts:17]: > Defines the

the what?

File: core/frontend/src/extension/extensions.ts
Comment (pmconne) [inline core/frontend/src/extension/extensions.ts:15]: Handles what?

**Comment by SpencerWBarnes** [comment]
> > * DisplayStyleSettings
> 
> Please don't delay 3.0 any further for this.
> 
> > * Replacement for `typeof` operation
> 
> Do you mean `instanceof`?

Thank you catching that Paul, updated the comment.

File: core/frontend/src/IModelApp.ts
Comment (calebmshafer) [reply core/frontend/src/IModelApp.ts:208]: Yea, good point. I'll switch it to onAfterStartup

File: core/frontend/src/extension/ExtensionRuntime.ts
Comment (aruniverse) [inline core/frontend/src/extension/ExtensionRuntime.ts:59]: @wgoehrig does it make sense to add this to other pkgs?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [inline core/frontend/src/extension/ExtensionAdmin.ts:56]: see comment above. This should be onAppStartup, even though even that is pretty ambiguous. Really something like onITwinJsInitialized might make more sense

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [inline core/frontend/src/extension/ExtensionAdmin.ts:81]: whats the reason for accepting an ambiguous function here as opposed to a typed manifest object that the consumer has fetched on their own (or the very least a type promise result)?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [inline core/frontend/src/extension/ExtensionAdmin.ts:109]: see comments in addBuildExtension. Is this necessary? Would it make more sense for the manifest to be type safe?

File: core/frontend/src/extension/ExtensionHost.ts
Comment (kckst8) [inline core/frontend/src/extension/ExtensionHost.ts:30]: do we want to open this can of worms? Is it not easier (and fair) for consumers to provide already localized strings (if they care to)

File: core/frontend/src/extension/ExtensionImpl.ts
Comment (kckst8) [inline core/frontend/src/extension/ExtensionImpl.ts:29]: I understand that this is V1 but this is very restrictive (basically can only add toolbar buttons to a single location)

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionAdmin.ts:37]: Yea, that makes more sense to me. I'll switch it around.

File: core/frontend/src/extension/ExtensionHost.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionHost.ts:30]: I'm debating on this one... The localization interface exposes a lot more than I'd like since it does have implications and side-effects.

Maybe we expose a limited API for `getLocalizedString`? That way an extension today could get them?  We'd need to quickly come up with a contribution point for adding localized strings though.

File: core/frontend/src/extension/ExtensionImpl.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionImpl.ts:29]: Yes, this is one of the first things we should tackle after the initial release. I'll make sure to put that on our list.

Comment (kckst8) [reply core/frontend/src/extension/ExtensionImpl.ts:29]: It's fine for v1. Getting the framework out is most important

Comment (aruniverse) [reply core/frontend/src/extension/ExtensionImpl.ts:29]: Its even more restrictive, its heavily tied with AppUI which idk is something we want to tie ourselves down to

I think we should think how extensions would plug into something not based on AppUi like display-test-app

Comment (pmconne) [reply core/frontend/src/extension/ExtensionImpl.ts:29]: @aruniverse Yes - this is exactly why display-test-app does not use our UI packages - to try to prevent implicit dependencies on them from creeping in elsewhere.

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts]: fixed

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts]: fixed

File: core/frontend/src/extension/extensions.ts
Comment (kckst8) [reply core/frontend/src/extension/extensions.ts]: fixed

File: core/frontend/src/extension/extensions.ts
Comment (kckst8) [reply core/frontend/src/extension/extensions.ts]: fixed

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts]: changed to onStart

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts]: updated to accept Promise<ExtensionManifest> and removed the extra parser function

File: core/frontend/src/extension/ExtensionHost.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionHost.ts]: removed for now

File: core/frontend/src/extension/ExtensionImpl.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionImpl.ts:29]: it's an "interesting" discussion. We certainly don't want to be tied to React, but the "provider" concept in general can probably be reused/re-purposed. The fact that we are allowing for extending UI that does not exist is not a trivial problem to solve

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionAdmin.ts]: removed

File: core/frontend/src/extension/ExtensionHost.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionHost.ts]: @kckst8 are you going to pick up removing this from the initial API?

Comment (kckst8) [reply core/frontend/src/extension/ExtensionHost.ts]: Already removed it from the host API and it doesn't appear to be exported for the extension runtime. Am I missing somewhere?

Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionHost.ts]: Nope, you're good. GitHub just hadn't updated for me...

File: core/frontend/src/extension/ExtensionImpl.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionImpl.ts:29]: True, I think we should definitely follow up on this and it's going to be a lengthy discussion. However, in terms of this PR are we okay with just hard-coding this how it is today?

As far as I know we did not expose anything explicitly AppUI related within the API and it's all under the hood to be potentially refactored later. If not, we should clear that up before merging.

File: core/frontend/src/extension/ExtensionRuntime.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionRuntime.ts:59]: Agree, I think it does for at the very least core-backend since you can never have multiple copies of that. However, not sure about the other ones in core. Their requirement to have a single one is driven by the core-frontend and core-backend pkgs.

Now some of the UI packages are another set that it may be helpful to do it for, mainly the framework ones.

File: core/frontend/src/extension/ExtensionHost.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionHost.ts:21]: We should definitely explore that to see if it's make things a bit easier. However, I think as we progress with the API we'll find we're adding things that are in addition to the existing iModelApp support or changing quite a bit so in that case it may be a bit too restrictive.

File: core/frontend/src/extension/ExtensionImpl.ts
Comment (kckst8) [reply core/frontend/src/extension/ExtensionImpl.ts:29]: this is good as is

File: core/frontend/src/extension/ExtensionRuntime.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionRuntime.ts:59]: @aruniverse added this to the backlog to at least tackle the core-backend.

File: core/frontend/src/extension/Extension.ts
Comment (calebmshafer) [inline core/frontend/src/extension/Extension.ts:17]: @kckst8 Do we need to expose this type?  It looks like we have an array for them at https://github.com/iTwin/itwinjs-core/pull/2743/files#diff-3ed6cce6e91b23ccf13b7553aad1615c9c0f3f6135fc8e0a59c3e5df116294c2R36.

File: core/frontend/src/extension/ExtensionHost.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionHost.ts:21]: made a note for us to revisit this and see if we'd want to switch. @aruniverse I'm going to close this one out.

File: core/frontend/src/extension/Extension.ts
Comment (kckst8) [reply core/frontend/src/extension/Extension.ts:17]: yeah we should decide on one for sure. I would prefer enums for this case. Thoughts?

Comment (calebmshafer) [reply core/frontend/src/extension/Extension.ts:17]: Agreed, I'd remove this type and export the Enum.  I'd mark it as `internal` as well since currently I don't see any reason an application would ever need it.

**Review by calebmshafer — APPROVED** [review]

**Comment by calebmshafer** [comment]
@pmconne mind taking a pass when you get a chance?

File: core/frontend/src/extension/Extension.ts
Comment (pmconne) [inline core/frontend/src/extension/Extension.ts:25]: Any uniqueness constraints? Difference from displayName?

File: core/frontend/src/extension/Extension.ts
Comment (pmconne) [inline core/frontend/src/extension/Extension.ts:30]: Does this follow semantic versioning, or otherwise expect a specifically-formatted string?

File: core/frontend/src/extension/Extension.ts
Comment (pmconne) [inline core/frontend/src/extension/Extension.ts:34]: Anything specific expected here? An email address or other contact info?

File: core/frontend/src/extension/Extension.ts
Comment (pmconne) [inline core/frontend/src/extension/Extension.ts:47]: and what?
If a BuildExtensionManifest is specifically about an extension loaded at build-time (I assume it is, docs don't say) then shouldn't the `module` property be required?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (pmconne) [inline core/frontend/src/extension/ExtensionAdmin.ts:13]: Where if anywhere does this block of documentation show up in the published doc site?

File: core/frontend/src/IModelApp.ts
Comment (pmconne) [inline core/frontend/src/IModelApp.ts:279]: Why is `IModelApp.extensionAdmin` `@alpha` when all the other APIs are `@beta`?

**Review by pmconne — COMMENTED** [review]
This is a major new feature. Where are the docs and example code? No tests?

File: core/frontend/src/extension/Extension.ts
Comment (kckst8) [reply core/frontend/src/extension/Extension.ts:25]: no constraints at this time. We aren't requiring uniqueness for build-time. That said, I think we may want to do so in the future (thinking mostly of runtime extensions). @calebmshafer @aruniverse  thoughts?

File: core/frontend/src/extension/Extension.ts
Comment (kckst8) [reply core/frontend/src/extension/Extension.ts:30]: this will be the version from package.json in most cases, so it "should" be semver but we haven't established that yet. I think its really up to the extension to just ensure versions are unique

File: core/frontend/src/extension/Extension.ts
Comment (kckst8) [reply core/frontend/src/extension/Extension.ts:34]: This will be from package.json in most cases, which would follow those [conventions](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#people-fields-author-contributors). So, if we start to care about the author (use in code), assuming that this is a string could be misleading. @calebmshafer @aruniverse I feel like we remove this for now. Thoughts?

File: core/frontend/src/extension/ExtensionAdmin.ts
Comment (calebmshafer) [reply core/frontend/src/extension/ExtensionAdmin.ts:13]: It does not today but it would be what we want to add as part of the learning documentation. It should be moved over there but those docs have yet to be written.

File: core/frontend/src/extension/Extension.ts
Comment (aruniverse) [reply core/frontend/src/extension/Extension.ts:25]: Based name and displayName off of what vscode does, see https://code.visualstudio.com/api/references/extension-manifest

for name i was thinking itd end up being the name defined in the pkg.json, and the displayName, if provided, would be the name shown when we eventually have an extension store/ui

i dont think the name or displayName need to be unique, we should have some guid on our end

File: core/frontend/src/extension/Extension.ts
Comment (kckst8) [reply core/frontend/src/extension/Extension.ts:47]: it should....and we actually aren't using this interface where we should be. I'll update

File: core/frontend/src/extension/Extension.ts
Comment (kckst8) [reply core/frontend/src/extension/Extension.ts]: removed

File: core/frontend/src/extension/Extension.ts
Comment (kckst8) [reply core/frontend/src/extension/Extension.ts]: updated and used appropriately

File: core/frontend/src/IModelApp.ts
Comment (kckst8) [reply core/frontend/src/IModelApp.ts]: updated to @beta

**Review by pmconne — APPROVED** [review]

**Pull request merged**
